src/test/java/com/google/sps/AuthServletTest.java

```diff
@@ -200,6 +219,136 @@ public void testAddUserToDatabasePresent() throws Exception {
     // Check whether the proper Entity and count were returned.
     Assert.assertEquals(1, listResults.size());
     Assert.assertEquals(EMAIL, listResults.get(0).getProperty(User.USER_EMAIL));
+
+    // Confirm that the Entity in the database matches the method return.
+    Assert.assertEquals(listResults.get(0), userEntityReturn);
+  }
+
+  @Test
+  public void testGetUserEntityFromEmailNotPresent() throws Exception {
+    // Run getUserEntityFromEmail(...), with the User not present in datastore.
+    Entity userEntityReturn = AuthServlet.getUserEntityFromEmail(EMAIL);
+
+    // Retrieve the datastore results.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query(User.USER);
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> listResults = results.asList(FetchOptions.Builder.withDefaults());
+
+    // Verify that the Entity is not in the database, and the null method return.
+    Assert.assertNull(userEntityReturn);
+    Assert.assertTrue(listResults.isEmpty());
+  }
+
+  @Test
+  public void testGetUserEntityFromEmailPresent() throws Exception {
+    // Add a User to the database.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity userEntity = new Entity(User.USER);
+    userEntity.setProperty(User.USER_EMAIL, EMAIL);
+    datastore.put(userEntity);
+
+    // Run getUserEntityFromEmail(...), with the User present in datastore.
+    Entity userEntityReturn = AuthServlet.getUserEntityFromEmail(EMAIL);
+
+    // Retrieve the datastore results.
+    Query query = new Query(User.USER);
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> listResults = results.asList(FetchOptions.Builder.withDefaults());
+
+    // Verify that the Entity is in the database, and the Entity method return.
+    Assert.assertEquals(EMAIL, userEntityReturn.getProperty(User.USER_EMAIL));
+    Assert.assertEquals(1, listResults.size());
+
+    // Confirm that the Entity in the database matches the method return.
+    Assert.assertEquals(listResults.get(0), userEntityReturn);
```

Comment:
is userEntityReturn also equal to userEntity created above? if so you can probably skip the part of this test which queries the db.

Reply:
They are actually slightly different; after you add it to datastore, the Entity gets a **key** and **id**; the `userEntity` object does not have these attributes. I'm testing that these are added and that the Entity objects are equal with the final line.

Reply:
👍